sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala

```diff
@@ -36,6 +38,9 @@ class RateStreamContinuousStream(rowsPerSecond: Long, numPartitions: Int) extend
 
   val perPartitionRate = rowsPerSecond.toDouble / numPartitions.toDouble
 
+  private[sql] val highestCommittedValue = new AtomicLong(Long.MinValue)
+  private[sql] val firstCommittedTime = new AtomicLong(Long.MinValue)
```

Comment:
I still have a doubt on the value of these. We usually don't add a variable like this just for testing.

Reply:
As I commented earlier, `creationTime` in this class above is one of fields just for testing. 

Unlike microbatch mode, it's not easy to track and capture the progress. That might be the reason we were relying on wait time. I totally agree the change is not beauty, but given rate source is for testing/benchmarking purpose, and end users don't deal with RateStreamContinuousStream directly, we might be just OK with it. 

If it's still a thing for us to be concerned, please let me know, and we can rollback the change and just add more seconds in wait time to make test be higher rate to be passed.

Reply:
The first approach is definitely more ugly than this. Just a question (without having a deep look) is there a metrics maybe which can be used? Or just add something like this?

Reply:
Stream query listener is not ready for continuous processing. #24537 is trying to address it, but seems like it's not ready. Actually the query listener mainly reports per batch to show changing numbers, and if we apply this to continuous processing, the report could be far behind if there's skew among partitions and epochs for partitions differ.

Reply:
Can you subclass this for testing and only add the new fields in the subclass?

Reply:
It would also require a new data source provider, as we just use it by the name of data source. 

The change would bring the test version of RateStreamProvider, and test version of RateStreamTable (maybe subclassing to deduplicate), and test version of RateStreamContinuousStream. I'd like to confirm whether it's OK to apply the change since the changeset is going to be bigger.

Reply:
I see, hmm, it's not instantiated directly.
I'm also uneasy about adding this just for a test; is there any way to avoid this while fixing the test, even if it means the test can't test as much?
I'm not super against this though. I'd mark it as visible for testing and comment of course.

Reply:
OK maybe I can try to find alternative approach first and stick with this only when I can't find one. I didn't do that since alternatives would be pretty less simpler, but I feel we would like to avoid the change like this.

Reply:
And I should admit that I just realized `creationTime` is not only used for testing. My bad. Sorry about the missing.